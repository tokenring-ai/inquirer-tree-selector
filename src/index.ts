import {
  createPrompt,
  makeTheme,
  type Status,
  useEffect,
  useKeypress,
  usePagination,
  usePrefix,
  useState
} from '@inquirer/core'
import { ANSI_HIDE_CURSOR, Status as LocalStatus } from '#consts'
import { baseTheme } from '#theme'
import type { PromptConfig } from '#types/config'
import type { Item } from '#types/item'
import type { StatusType } from '#types/status'
import type { PromptTheme, RenderContext } from '#types/theme'
import {
  isDownKey,
  isEnterKey,
  isEscapeKey,
  isLeftKey,
  isPageDown,
  isPageUp,
  isRightKey,
  isSpaceKey,
  isUpKey
} from '#utils/key'

// Multiple selection enabled, cancellation disabled
export function treeSelector(
  config: PromptConfig & { multiple: true; allowCancel?: false }
): Promise<string[]>

// Multiple selection enabled, cancellation enabled
export function treeSelector(
  config: PromptConfig & { multiple: true; allowCancel: true }
): Promise<string[] | null>

// Single selection, cancellation disabled (existing)
export function treeSelector(
  config: PromptConfig & { multiple?: false; allowCancel?: false }
): Promise<string>

// Single selection, cancellation enabled (existing)
export function treeSelector(
  config: PromptConfig & { multiple?: false; allowCancel: true }
): Promise<string | null>

// Main implementation signature
export function treeSelector(
  config: PromptConfig
): Promise<string | string[] | null> {
  return createPrompt<string | string[] | null, PromptConfig>(
    (config, done) => {
      const {
        pageSize = 10,
        loop = false,
        allowCancel = false,
        cancelText = 'Canceled.',
        emptyText = 'No items available.',
        multiple = false,
        initialSelection,
        tree
      } = config

      const [status, setStatus] = useState<Status>(LocalStatus.Idle as Status)
      const theme = makeTheme<PromptTheme>(baseTheme, config.theme)
      const prefix = usePrefix({ status, theme })
      const [selectedItems, setSelectedItems] = useState<string[]>(
        Array.isArray(initialSelection) ? initialSelection : [initialSelection]
      )

      const [itemStack, setItemStack] = useState<Item[]>([tree])

      const currentItem = itemStack[itemStack.length - 1]

      const [children, setChildren] = useState<Item[] | undefined>()

      useEffect(() => {
        if (currentItem.children instanceof Function) {
          setChildren(undefined)
          Promise.resolve(currentItem.children()).then(children =>
            setChildren(children ?? [])
          )
        } else {
          process.nextTick(() =>
            setChildren((currentItem.children as Item[]) ?? [])
          )
        }
      }, [currentItem, currentItem?.children])

      const [active, setActive] = useState(0)

      const activeItem = children != null ? children[active] : undefined

      useKeypress((key, rl) => {
        if (isEnterKey(key)) {
          if (multiple) {
            // In multiple mode, Enter completes the selection with all selected items
            setStatus(LocalStatus.Done as Status)
            done(selectedItems)
          } else {
            if (activeItem?.value == null) {
              return
            }

            // In single mode, Enter selects the current item
            setStatus(LocalStatus.Done as Status)
            done(activeItem.value)
          }
        } else if (isSpaceKey(key)) {
          // Space now handles selection (toggle in multiple mode, select in single mode)
          if (activeItem?.value == null) {
            return
          }

          if (multiple) {
            const idx = selectedItems.indexOf(activeItem.value)
            if (idx !== -1) {
              setSelectedItems([
                ...selectedItems.slice(0, idx),
                ...selectedItems.slice(idx + 1)
              ])
            } else {
              setSelectedItems([...selectedItems, activeItem.value])
            }
          } else {
            // In single mode, space selects the item (same as Enter)
            setStatus(LocalStatus.Done as Status)
            done(activeItem.value)
          }
        } else if (isRightKey(key)) {
          // Right arrow navigates into children - only if the item has children
          if (activeItem?.children) {
            setItemStack([...itemStack, activeItem])
          }
        } else if (isLeftKey(key) && itemStack.length > 1) {
          setItemStack(itemStack.slice(0, -1))
        } else if (isEscapeKey(key) && allowCancel) {
          setStatus(LocalStatus.Canceled as Status)
          done(null)
        } else if (
          isUpKey(key) ||
          isDownKey(key) ||
          isPageUp(key) ||
          isPageDown(key)
        ) {
          rl.clearLine(0)

          let newOffset = isUpKey(key)
            ? active - 1
            : isDownKey(key)
              ? active + 1
              : isPageUp(key)
                ? active - pageSize
                : active + pageSize

          if (loop) {
            newOffset = children?.length ?? 0 % newOffset
          } else {
            newOffset = Math.min(
              Math.max(newOffset, 0),
              (children?.length ?? 0) - 1
            )
          }

          setActive(newOffset)
        }
      })

      const page = usePagination({
        items: children ?? [],
        active,
        renderItem: ({ item, index, isActive }) => {
          return theme.renderItem(item, {
            items: children ?? [],
            loop,
            index,
            isActive,
            isSelected:
              item.value != null && selectedItems.includes(item.value),
            multiple
          })
        },
        pageSize,
        loop
      })

      const message = theme.style.message(config.message, status)

      if (status === (LocalStatus.Canceled as Status)) {
        return `${prefix} ${message} ${theme.style.cancelText(cancelText)}`
      }

      if (activeItem && status === (LocalStatus.Done as Status)) {
        if (multiple) {
          // In multiple mode, show the count of selected items
          const count = selectedItems.length
          const itemText = count === 1 ? 'item' : 'items'
          return `${prefix} ${message} ${theme.style.answer(`${count} ${itemText} selected`)}`
        }
        // In single mode, show the selected item name
        return `${prefix} ${message} ${theme.style.answer(activeItem.name)}`
      }

      const helpTop = theme.style.help(theme.help.top(allowCancel, multiple))
      const content =
        children == null
          ? 'Loading...'
          : !page.length
            ? theme.style.emptyText(emptyText)
            : page

      return `${prefix} ${message} ${helpTop}\n${content}${ANSI_HIDE_CURSOR}`
    }
  )(config)
}

export { LocalStatus as Status }

export type { StatusType, PromptConfig, Item, PromptTheme, RenderContext }
