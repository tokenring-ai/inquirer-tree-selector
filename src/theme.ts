import figures from '@inquirer/figures'
import chalk from 'chalk'
import type { Item } from '#types/item'
import type { StatusType } from '#types/status'
import type { PromptTheme, RenderContext } from '#types/theme'

export const baseTheme: PromptTheme = {
  prefix: {
    idle: chalk.cyan('?'),
    done: chalk.green(figures.tick),
    canceled: chalk.red(figures.cross)
  },
  style: {
    active: (text: string) => chalk.cyan(text),
    selected: (text: string) => chalk.green(`${text} ${figures.tick}`),
    unselected: (text: string) => text,
    cancelText: (text: string) => chalk.red(text),
    emptyText: (text: string) => chalk.red(text),
    group: (text: string) => chalk.yellowBright(text),
    item: (text: string) => text,
    message: (text: string, _status: StatusType) => chalk.bold(text),
    help: (text: string) => chalk.italic.gray(text),
    answer: (text: string) => chalk.cyan(text)
  },
  hierarchySymbols: {
    branch: figures.lineUpDownRight + figures.line,
    leaf: figures.lineUpRight + figures.line
  },
  help: {
    top: (allowCancel: boolean, multiple?: boolean) =>
      `(Press ${figures.arrowUp + figures.arrowDown} to navigate, ${figures.arrowLeft + figures.arrowRight} to navigate tree${
        multiple ? ', <space> to toggle selection' : ''
      }, <enter> to ${multiple ? 'confirm' : 'select'}${allowCancel ? ', <esc> or q to cancel' : ''})`,
    item: multiple =>
      multiple
        ? '(Press <space> to toggle selection)'
        : '(Press <enter> to select)'
  },
  renderItem(item: Item, context: RenderContext) {
    const isLast = context.index === context.items.length - 1
    let line =
      isLast && !context.loop
        ? this.hierarchySymbols.leaf
        : this.hierarchySymbols.branch

    if (context.isSelected) {
      line += this.style.selected(item.name)
    } else {
      if (context.isActive) {
        line += this.style.active(item.name)
      } else if (context.isSelected) {
        line += this.style.selected(item.name)
      } else if (item.children) {
        const hasEnabledItems = context.selectedItems.some(
          selectedItem =>
            item.value != null && selectedItem.startsWith(item.value)
        )
        if (hasEnabledItems) {
          line += this.style.selected(item.name)
        } else {
          line += this.style.group(item.name)
        }
      } else if (isLast && !context.loop) {
        line += this.style.unselected(item.name)
      } else {
        line += this.style.unselected(item.name)
      }
      line += '  '
    }

    if (context.isActive) {
      const helpMessage = this.help.item(context.multiple)
      line += ` ${this.style.help(helpMessage)}`
    }

    return line
  }
}
