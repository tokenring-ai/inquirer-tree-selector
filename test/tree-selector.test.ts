import { describe, it, expect, vi } from 'vitest'
import { treeSelector } from '../src'
import { render } from '@inquirer/testing'

const tree = [
  {
    name: 'A',
    value: 'a',
    children: () => Promise.resolve([
      { name: 'A1', value: 'a1' },
      { name: 'A2', value: 'a2' },
    ]),
  },
  {
    name: 'B',
    value: 'b',
    children: () => Promise.resolve([
      { name: 'B1', value: 'b1' },
      { name: 'B2', value: 'b2' },
    ]),
  },
]

describe('treeSelector', () => {
  it('should select a single item', async () => {
    const { answer, events, getScreen } = await render(
      treeSelector,
      {
        message: 'Select an item',
        tree,
      }
    )

    events.keypress('down')
    await Promise.resolve()
    events.keypress('enter')
    await Promise.resolve()

    await expect(answer).resolves.toEqual('a1')
    expect(getScreen()).toMatchSnapshot()
  })

  it('should select multiple items', async () => {
    const { answer, events, getScreen } = await render(
      treeSelector,
      {
        message: 'Select an item',
        tree,
        multiple: true,
      }
    )

    events.keypress('down')
    await Promise.resolve()
    events.keypress('space')
    await Promise.resolve()
    events.keypress('down')
    await Promise.resolve()
    events.keypress('space')
    await Promise.resolve()
    events.keypress('enter')
    await Promise.resolve()

    await expect(answer).resolves.toEqual(['a1', 'a2'])
    expect(getScreen()).toMatchSnapshot()
  })

  it('should show enabled leafs', async () => {
    const { answer, events, getScreen } = await render(
      treeSelector,
      {
        message: 'Select an item',
        tree,
        multiple: true,
        initialSelection: ['a1'],
      }
    )

    events.keypress('enter')
    await Promise.resolve()

    await expect(answer).resolves.toEqual(['a1'])
    expect(getScreen()).toMatchSnapshot()
  })
})
