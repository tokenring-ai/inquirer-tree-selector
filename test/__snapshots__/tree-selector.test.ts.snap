// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`treeSelector > should render the prompt 1`] = `""`;
